### PR TITLE
[ESP-12] Store the the EmailID from the SMTP response

### PR DESCRIPTION
**Pull request by saikambaiyyagari**
**Description:**
The PR adds a code path to store the ULID EmailID value from the SMTP response to a `Mail::Message` header - `email_id`. 
The regex to detect the presence of ULID value conforms with spec - https://github.com/ulid/spec#encoding : 
- Canonically encoded as a 26 character string
- Excludes the letters I, L, O, and U to avoid confusion and abuse.

When the SMTP response (from Postfix) doesn't return a ULID value in the message, `Mail::Message` will not contain a header `email_id`.  We have added a few additional tests to confirm if this behavior and cover some edge case scenarios.

**JIRA:**
https://zendesk.atlassian.net/browse/ESP-12

**Risk**
Low - The code path to store EmailID from SMTP response if present. We have some additional precaution checks - `response.respond_to?(:message) ` and `response&.message.nil?` to handle edge case scenarios.

### CODE REVIEW

**Comment by saikambaiyyagari** [comment]
PR to address migration from Travis to GHA - https://github.com/zendesk/action_mailer-logged_smtp_delivery/pull/12

File: lib/action_mailer/logged_smtp_delivery.rb [hunk 1]
```diff
@@ -2,6 +2,8 @@
 require 'mail/network/delivery_methods/smtp'
 
 class ActionMailer::LoggedSMTPDelivery < Mail::SMTP
+  ULID_PATTERN = /[ABCDEFGHJKMNPQRSTVWXYZ0-9]{26}/.freeze
```
Comment (manvil) [inline lib/action_mailer/logged_smtp_delivery.rb:5]:
```suggestion
  ULID_PATTERN = /[A-Z0-9]{26}/.freeze
```
^^ would this work?

**Review by manvil — CHANGES_REQUESTED** [review]
One minor comment, otherwise looks good

File: lib/action_mailer/logged_smtp_delivery.rb (hunk 1 above)
Comment (saikambaiyyagari) [reply lib/action_mailer/logged_smtp_delivery.rb:5]: We had this `/[A-Z0-9]{26}/` regex initially but according to the spec - [ULID](https://github.com/ulid/spec#encoding ) value `excludes the letters I, L, O, and U to avoid confusion and abuse.` Hence we updated the regex it to `/[ABCDEFGHJKMNPQRSTVWXYZ0-9]{26}/` 🙏

Comment (manvil) [reply lib/action_mailer/logged_smtp_delivery.rb:5]:
```suggestion
ULID_PATTERN = /[A-HJ-KMNV-Z0-9]{26}/.freeze
```
Would this be better? Also it would be to put the comment that we are omitting letters and why we are doing it

**Comment by daphsta** [comment]
could we add a test for an invalid ULID value as well? one that contains I,L,O,U

File: lib/action_mailer/logged_smtp_delivery.rb (hunk 1 above)
Comment (saikambaiyyagari) [reply lib/action_mailer/logged_smtp_delivery.rb:5]: Sure, I will add the comment about omitting the letters

Comment (saikambaiyyagari) [reply lib/action_mailer/logged_smtp_delivery.rb:5]: How about this regex? /[A-HJKMNV-Z0-9]{26}/.freeze

**Comment by saikambaiyyagari** [comment]
> could we add a test for an invalid ULID value as well? one that contains I,L,O,U
We have added a test to check the invalid value

**Review by manvil — APPROVED** [review]

File: lib/action_mailer/logged_smtp_delivery.rb [hunk 2]
```diff
@@ -2,6 +2,11 @@
 require 'mail/network/delivery_methods/smtp'
 
 class ActionMailer::LoggedSMTPDelivery < Mail::SMTP
+  # The ULID pattern conforms with the spec - https://github.com/ulid/spec#encoding
+  # a. Canonically encoded as a 26 character string
+  # b. Excludes the letters I, L, O, and U to avoid confusion and abuse.
+  ULID_PATTERN = /[A-HJKMNP-TV-Z0-9]{26}/.freeze
```
Comment (daphsta) [inline lib/action_mailer/logged_smtp_delivery.rb:8]: Just curious if negative lookahead might make this more readable. Something like 
/(?!.*[ILOU])[A-Z0-9]{26}/

Comment (saikambaiyyagari) [reply lib/action_mailer/logged_smtp_delivery.rb:8]: This looks more readable and clear. Thanks for suggesting it.
